kvserver: circuit-break requests to unavailable ranges

tbg opened the pull request:
Fixes #33007.
Closes #61311.

This PR uses the circuit breaker package introduced in #73641 to address
issue #33007: When a replica becomes unavailable, it should eagerly
refuse traffic that it believes would simply hang.

Concretely, every request to the Replica gets a cancelable context that
is sensitive to the circuit breaker tripping (relying on context
cancellation makes sure we don't end up with a second channel that needs
to be plumbed to everyone and their dog; Go makes it really difficult to
join two cancellation chains); if the breaker is tripped when the
request arrives, it is refused right away. In either case, the outgoing
error is augmented to carry information about the tripped breaker. This
isn't 100% trivial, since retreating from in-flight replication
typically causes an `AmbiguousResultError`, and while we could avoid it
in some cases we can't in all. A similar problem occurs when the
canceled request was waiting on a lease, in which case the result is a
NotLeaseholderError.

For any request that made it in, if it enters replication but does not
manage to succeed within `base.SlowRequestThreshold` (15s at time of
writing), the breaker is tripped, cancelling all inflight and future
requests until the breaker heals. Perhaps surprisingly, the existing
"slowness" detection (the informational "have been waiting ... for
proposal" warning in `executeWriteBatch`) was moved deeper into
replication (`refreshProposalsLocked`), where it now trips the breaker.
This has the advantage of providing a unified path for lease requests
(which don't go through `executeWriteBatch`) and pipelined writes (which
return before waiting on the inflight replication process). To make this
work, we need to pick a little fight with how leases are (not) refreshed
(#74711) and we need to remember the ticks at which a proposal was first
inserted (despite potential reproposals).

A tripped breaker deploys a background probe which sends a
`ProbeRequest` (introduced in #72972) to determine when to heal; this is
roughly the case whenever replicating a command through Raft is possible
for the Replica, either by appending to the log as the Raft leader, or
by forwarding to the Raft leader. A tiny bit of special casing allows
requests made by the probe to bypass the breaker.

As of this PR, all of this is opt-in via the
`kv.replica_circuit_breakers.enabled` cluster setting while we determine
which of the many follow-up items to target for 22.1 (see #74705).  For
example, the breaker-sensitive context cancelation is a toy
implementation that comes with too much of a performance overhead (one
extra goroutine per request); #74707 will address that.

Other things not done here that we certainly want in the 22.1 release
are UI work (#74713) and metrics (#74505).

The release note is deferred to #74705 (where breakers are turned on).

Release note: None

tbg committed:
kv: circuit-break requests to unavailable replicas

Fixes #33007.
Closes #61311.

This PR uses the circuit breaker package introduced in #73641 to address
issue #33007: When a replica becomes unavailable, it should eagerly
refuse traffic that it believes would simply hang.

Concretely, every request to the Replica gets a cancelable context that
is sensitive to the circuit breaker tripping (relying on context
cancellation makes sure we don't end up with a second channel that needs
to be plumbed to everyone and their dog; Go makes it really difficult to
join two cancellation chains); if the breaker is tripped when the
request arrives, it is refused right away. In either case, the outgoing
error is augmented to carry information about the tripped breaker. This
isn't 100% trivial, since retreating from in-flight replication
typically causes an `AmbiguousResultError`, and while we could avoid it
in some cases we can't in all. A similar problem occurs when the
canceled request was waiting on a lease, in which case the result is a
NotLeaseholderError.

For any request that made it in, if it enters replication but does not
manage to succeed within the timeout specified by the
`kv.replica_circuit_breaker.slow_replication_threshold` cluster setting,
the breaker is tripped, cancelling all inflight and future requests
until the breaker heals. Perhaps surprisingly, the existing "slowness"
detection (the informational "have been waiting ... for proposal"
warning in `executeWriteBatch`) was moved deeper into replication
(`refreshProposalsLocked`), where it now trips the breaker.  This has
the advantage of providing a unified path for lease requests (which
don't go through `executeWriteBatch`) and pipelined writes (which return
before waiting on the inflight replication process). To make this work,
we need to pick a little fight with how leases are (not) refreshed
(#74711) and we need to remember the ticks at which a proposal was first
inserted (despite potential reproposals).

Perhaps surprisingly, when the breaker is tripped, *all* traffic to
the Replica gets the fail-fast behavior, not just mutations. This is
because even though a read may look good to serve based on the lease,
we would also need to check for latches, and in particular we would
need to fail-fast if there is a transitive dependency on any write
(since that write is presumably stuck). This is not trivial and so
we don't do it in this first iteration (see #74799).

A tripped breaker deploys a background probe which sends a
`ProbeRequest` (introduced in #72972) to determine when to heal; this is
roughly the case whenever replicating a command through Raft is possible
for the Replica, either by appending to the log as the Raft leader, or
by forwarding to the Raft leader. A tiny bit of special casing allows
requests made by the probe to bypass the breaker.

As of this PR, the cluster setting defaults to zero (disabling the
entire mechanism) until some necessary follow-up items have been
addressed (see #74705). For example, the breaker-sensitive context
cancelation is a toy implementation that comes with too much of a
performance overhead (one extra goroutine per request); #74707 will
address that.  Other things not done here that we certainly want in the
22.1 release are UI work (#74713) and metrics (#74505).

The release note is deferred to #74705 (where breakers are turned on).

Release note: None

touchie

tbg commented:
@cockroachdb/repl-prs this is still a draft but it is ready for a high-level review. There is a lot going on and I can easily pull 5+ pieces of this into separate PRs and am planning to do so. However, before I start that I would like to get both of you on board with the basic approach. I think we've sufficiently discussed the circuit breaker API and are happy with it; I will rewrite the internals to avoid all of the unsafe stuff so ignore the internals but give the API another look and voice any larger concerns. For the NoopWrite, there is an alternative where we don't bypass the lease. I think bypassing the lease has worked out nicely and keeps `NoopWrite` simple and possibly reusable, but on the other hand the probe may pass but for some unknown reason a lease may never be acquired. My preference is to keep NoopWrite as is but make the probe use two stages, where it issues the noop write first and a lease-requiring write second but possibly this is a bad idea and I need to sleep over it and talk it through.

erikgrinaker reviewed (commented):
> Reviewed 5 of 5 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71806)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker and @tbg)

___
*[pkg/util/circuit/circuitbreaker.go, line 28 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71806#-Mo3s0Th8t9zXmDBhS3d:-Mo3s0Th8t9zXmDBhS3e:b-5e6513) ([raw file](https://github.com/cockroachdb/cockroach/blob/fac33a5757d87494e49f52dc3e1c26065cc18c09/pkg/util/circuit/circuitbreaker.go#L28)):*
> ```Go
> // from Breaker.Err(), i.e. `errors.Is(err, ErrBreakerOpen()) can be
> // used to check whether an error originated from some Breaker.
> func ErrBreakerOpen() error {
> ```

not: why not simply export the variable?
___
*[pkg/util/circuit/circuitbreaker.go, line 56 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71806#-Mo3swPQ8OQwyUh0Ohqv:-Mo3sYCTCO0idYRIWvEs:bg5m4qg) ([raw file](https://github.com/cockroachdb/cockroach/blob/fac33a5757d87494e49f52dc3e1c26065cc18c09/pkg/util/circuit/circuitbreaker.go#L56)):*
> ```Go
> 	Name         string
> 	ShouldTrip   func(err error) error
> 	AsyncProbe   func(report func(error), done func())
> ```

Why does this need to take a `done` closure? Since the probe is expected to keep running, can't we just clean up when the closure returns?

Ah, I suppose it's because it's the probe's responsibility to spawn the async task. Is there a reason for that, i.e. why can't the prober spawn the task? The only reason I can see is for the caller to pass in its context, but that seems like an antipattern anyway?


<!-- Sent from Reviewable.io -->

tbg commented:
Spent some time on this again today. While putting together an end-to-end test (lose quorum, try to write and see the failfast) I realized that we do need to make all parts of the read/write pipeline aware of the circuit breaker to avoid getting stuck. I knew that we needed lease proposals to be sensitive to the breaker (so that you wouldn't get stuck trying to acquire a lease over and over when there wasn't a valid lease to begin with) but I wasn't aware of a much more basic problem: that when one command fails fast after having proposed a command, the latches stick around. They have to; but this means that the next command will get stuck on the latches. So we need to short-circuit the latch acquisition as well when the breaker trips. Of course we don't want to plumb the circuit breaker everywhere, you'd really want to rely on `context.Context` for that sort of thing. I've hacked something together in the current version of the PR (as pushed right now) which I can polish to something that should get the job done for v1. Down the road, it would be interesting to explore https://github.com/cockroachdb/cockroach/pull/66884 as an ingredient here but I don't want this to become a prerequisite.

erikgrinaker commented on pkg/kv/kvserver/client_replica_circuit_breaker_test.go:
```diff
@@ -119,6 +120,15 @@ func TestReplicaCircuitBreaker(t *testing.T) {
 		tc.RequireIsNotLeaseholderError(t, tc.Write(n2))
 	})
+
+	runCircuitBreakerTest(t, "leaseholder-unavailable", func(t *testing.T, ctx context.Context, tc *circuitBreakerTest) {
```
This should probably go into the next commit, "wip: add leaseholder-unavailable test".

erikgrinaker commented:
>when one command fails fast after having proposed a command, the latches stick around. They have to; but this means that the next command will get stuck on the latches. So we need to short-circuit the latch acquisition as well when the breaker trips. Of course we don't want to plumb the circuit breaker everywhere, you'd really want to rely on `context.Context` for that sort of thing

Ugh, this is somewhat annoying. Conceptually, cancelling the context in `Send()` when the breaker trips makes sense -- it's essentially what we want to happen. But the bookkeeping gets a bit unwieldy, as you point out. It'd be nice if we could inject a cancel trigger into the context somehow, without needing to keep track of the individual cancel functions.

I'll have to remind myself of how the latch guards are managed, but if the latches turn out to be the only problem here, is it terrible to come up with a solution at/below the Raft level that causes us to abandon proposals/commands and release latches without having to do context management higher up? Although I realize that the context approach generalizes better across failure modes.

tbg commented:
Discussed this with Erik yesterday. Despite the potential perf overhead we decided that it was an important property that requests get canceled reliably on breaker tripping. We are abandoning our initial ambition to let reads that can be served succeed: this is just hard to build correctly. The main reason for reads to get stuck (no lease) is handled adequately by only circuit breaking on writes (lease acquisition is a write), but it's trickier if there is a stuck in-flight write that then prevents reads from getting their latches. We'd basically have to tell the latch manager to fail-fast a read but only if it is transitively waiting on a write's latch (write latches have to stick around on unavailable ranges, even if the client has long been fail-fasted). The "transitively" here is not something the latch manager is even concerned with today, see https://github.com/cockroachdb/cockroach/pull/73916 for a related discussion.

so:

- early in Replica.Send, check the breaker, make context cancelable and stash the cancel func somewhere on `Replica`
- on break, have the probe cancel everyone.

tbg commented:
Preparing a demo now (demo.sh in PR) and debugging some issue that I think is unrelated to the circuit breakers. Essentially I'm running kv0 against a cluster that has lost quorum for some kv ranges (but there is always at least one replica left), and sometimes the workload will get stuck instead of going into a pure fail-fast regime. There aren't any stacks in DistSender or on the kvserver side when this happens, so I think I'm tickling an unrelated bug. Further evidence of this - the problem persists when I restart the down nodes. Documenting troubleshooting below.

I have n1-n5, where n1 and n2 are down. Goroutine stacks for n3-n5 have all `pgwire` serving conns in `(*readTimeoutConn).Read`; so the below comes back empty (first negative grep is for the Read call, second restricts to goroutines serving pgwire through `(*conn).serveImpl`:

```
cat goroutine* | ./scripts/sgrep '!/conn.go:362/ && /conn.go:496/'
(empty)
```

In other words, all pgwire goroutines on the server are waiting for client input.

Similarly, all connExecutors attached to these conns are in `(*StmtBuf).CurCmd`:

```
cat goroutine* | ./scripts/sgrep '/onnExecutor/ && !/conn_io.go:441/'
(empty)
```

So they are waiting to be fed the next command as well (i.e. it's not like they have one going right now):

https://github.com/cockroachdb/cockroach/blob/4ef90c25dc9914e3e7254b18ae06e02fc9c0bc53/pkg/sql/conn_io.go#L440-L441

Here's the stderr of workload including stack trace (I killed it using `SIGQUIT`): https://gist.github.com/tbg/409fac0316acb0764765f0e37116b7fd

Notably, there are 32 workers (matching the default concurrency flag on my machine) and 12 of them are stuck acquiring a connection from the pool, whereas the other 20 are stuck... doing something else, see below.

```
$ cat errors.txt | scripts/sgrep '/StmtHandle/' | grep -c 'goroutine '
32
$ cat errors.txt | scripts/sgrep '/StmtHandle/ && /..Conn..Exec/' | grep -c 'goroutine '
20
$ cat errors.txt | scripts/sgrep '/StmtHandle/ && /..Pool..Acquire/' | grep -c 'goroutine '
12
```

The 20 are all here:

https://github.com/cockroachdb/vendored/blob/master/github.com/jackc/pgconn/pgconn.go#L1437

with identical stacks (https://gist.github.com/tbg/77de5731a3971ff5235439b30259f67f). In particular, they're also waiting to receive a message from the server (recall that the server is also waiting for messages from the client).

So I think there's something wrong here, probably involving errors being returned, and perhaps prepared statements?

I tried to reproduce with this branch:

https://github.com/cockroachdb/cockroach/compare/master...tbg:stuck-workload-repro?expand=1

but it looks to be chugging along nicely, and at least six minutes in I am not seeing an obvious starvation of worker goroutines. In particular, nobody is sitting on `(*Pool).Acquire` for a long time.

tbg commented:
Recording for later since I'm cleaning up the branch, [demo.sh](https://gist.github.com/d22b8b4650aa08ad2a827bf7080d8aa9)

tbg commented:
Something I need to look into tomorrow: when I run the demo, and reduce duration to zero, workload stops receiving errors, as expected. But when I then re-enable the breakers, the errors don't start flowing again.

tbg committed:
echotest: add testing helper

We have a few tests that record the output of formatting operations.
Historically this meant pasting the expected output into source code and
updating it whenenever necessary.

A pattern I've been using over the last year or so is to use datadriven
as a convenience wrapper for this, to a) avoid polluting the test code
files and b) to be able to use the `-rewrite` flag for convenience.

Personally I think this deserves its own helper, as this is a pattern
worth spreading. I have another use for it in #71806, and there are
a few more across the codebase that I plan to update as I run into
them.

Release note: None

erikgrinaker commented on pkg/kv/kvserver/client_replica_circuit_breaker_test.go:
```diff
@@ -0,0 +1,566 @@
+		// and none of the latches are taken), but since it is hard to determine
+		// that upfront we currently fail all reads as well.
+		tc.RequireIsBreakerOpen(t, tc.Read(n1))
+		tc.RequireIsBreakerOpen(t, tc.Write(n1))
```
I wonder if we'd want to add a latency assertion for these operations, to make sure we actually fail fast. These can be pretty annoying to test reliably, but it's an important property of the breaker that requests don't appear to hang.

erikgrinaker commented on pkg/kv/kvserver/client_replica_circuit_breaker_test.go:
```diff
@@ -0,0 +1,566 @@
+	// shut down n2 and expire the lease. n1 will be a non-leaseholder without
+	// quorum, and requests to it should trip the circuit breaker. This is an
+	// interesting test case internally because here, the request that trips the
+	// breaker is the slow lease request, and not the tests's actual write. Since
```
nit: tests's -> test's

erikgrinaker commented on pkg/kv/kvserver/replica.go:
```diff
@@ -264,6 +264,8 @@ type Replica struct {
 	// miss out on anything.
 	raftCtx context.Context
 
+	breaker *replicaCircuitBreaker
```
Needs a comment.

erikgrinaker commented on pkg/kv/kvserver/replica_circuit_breaker.go:
```diff
@@ -0,0 +1,183 @@
+	"kv.replica_circuit_breaker.slow_replication_threshold",
+	"duration after which slow proposals trip the per-Replica circuit breaker (zero duration disables breakers)",
+	defaultReplicaCircuitBreakerSlowReplicationThreshold,
+	func(d time.Duration) error {
```
`settings.NonNegativeDuration`

erikgrinaker commented on pkg/kv/kvserver/replica_circuit_breaker.go:
nit: consider doing this at the start of the function, since it negates the entire function.

erikgrinaker commented on pkg/kv/kvserver/replica_circuit_breaker.go:
nit: unnecessary `Sprintf()`, a literal string will autocast to `redact.RedactableString`.

erikgrinaker commented on pkg/kv/kvserver/replica_circuit_breaker.go:
nit: the arguments here all fit on one line, which reads better.

erikgrinaker commented on pkg/kv/kvserver/replica_raft.go:
```diff
@@ -1092,9 +1107,25 @@ func (r *Replica) refreshProposalsLocked(
 	}
 
+	var maxInflightDurationReq *roachpb.BatchRequest
+	var maxInflightDuration time.Duration
```
nit: consider calling these e.g. `maxSlowProposalDuration` or something, since they're conditional on the proposals being considered slow.

erikgrinaker commented on pkg/kv/kvserver/replica_raft.go:
```diff
@@ -1092,9 +1107,25 @@ func (r *Replica) refreshProposalsLocked(
 	for _, p := range r.mu.proposals {
-		if p.command.MaxLeaseIndex == 0 {
+		slowReplicationThreshold, ok := r.slowReplicationThreshold(p.Request)
+		inflightDuration := r.store.cfg.RaftTickInterval * time.Duration(r.mu.ticks-p.createdAtTicks)
```
Do we really want to rely on ticks here? Since ticks are driven by the scheduler, they can drift significantly from real time if the scheduler is slow (see #66686).

erikgrinaker replied:
I think it's worth making a note and revisiting later. If the scheduler is overwhelmed it could cause ranges to essentially be unavailable but the breakers would be very slow to trip. But no need to block this PR on that.

erikgrinaker commented on pkg/kv/kvserver/replica_raft.go:
```diff
@@ -1092,9 +1107,25 @@ func (r *Replica) refreshProposalsLocked(
+		// do about #74711. If leases are finished instead of reproposed, they can't
+		// ever trigger the breaker, which is bad as there usually isn't anything
+		// else around that will.
+		if p.command.MaxLeaseIndex == 0 && !r.breaker.enabled() {
```
Do we plan on resolving this hack before 22.1?

erikgrinaker commented on pkg/kv/kvserver/replica_range_lease.go:
```diff
@@ -1216,6 +1221,11 @@ func (r *Replica) redirectOnOrAcquireLeaseForRequest(
+			// it guards the actual lease proposal, which is a proposal and which
+			// thus already has a "slow message" trigger attached to it.
+			//
+			// TODO(during review): decide whether to remove this right now.
```
One argument for keeping this is that the slow proposal logging will only log the oldest request, and so it can mask the lease acquisition. We may want to add conditional slow request logging for significant system proposals such as lease acquisitions.

Given how delicate the lease acquisition stuff is, we may want to leave this in for another major release cycle, and tear it out once we're more comfortable with the new breakers. That will also retain this logging here if we either end up disabling breakers by default, or users disable it explicitly.

erikgrinaker commented on pkg/kv/kvserver/replica_raft.go:
```diff
@@ -1141,6 +1172,30 @@ func (r *Replica) refreshProposalsLocked(
+		// breaker (since the breaker needs access to this information at will to
+		// power the probe anyway). Over time, we anticipate there being multiple
+		// mechanisms which trip the breaker.
+		r.breaker.TripAsync()
```
Since this is async, could we end up logging the above multiple times if we're called again before the breaker actually trips?

erikgrinaker replied:
Yeah, 3s seems reasonably unlikely.

erikgrinaker commented on pkg/kv/kvserver/replica_send.go:
```diff
@@ -126,6 +126,71 @@ func (r *Replica) sendWithRangeID(
+				cancel()
+			}
+		}) != nil {
+			cancel()
```
nit: this is going to get called by the `defer` anyway.

erikgrinaker commented on pkg/kv/kvserver/replica_send.go:
```diff
@@ -126,6 +126,71 @@ func (r *Replica) sendWithRangeID(
 		return nil, roachpb.NewError(err)
 	}
 
+	// TODO(during review): try to move some of this code into a method to keep
```
Yes please.

erikgrinaker commented on pkg/kv/kvserver/replica_write.go:
```diff
@@ -345,52 +316,52 @@ func (r *Replica) executeWriteBatch(
+	return err
+}
+
+func (r *Replica) rangeUnavailableError() error {
```
nit: consider moving this elsewhere (preferably into `replica_circuit_breaker.go`) -- it is no longer relevant for this file.

erikgrinaker commented:
This turned out really nice, it was a pleasure to review.

I'll hold off on an actual approval until it's marked as ready, but this is essentially good to go I think.

erikgrinaker reviewed (commented):
This turned out really nice, it was a pleasure to review.

I'll hold off on an actual approval until it's marked as ready, but this is essentially good to go I think.

tbg reviewed (commented):
Can't comment on some of your review comments any more, so here you go:

>  We should call this out in the commit message, along with the rationale -- it's fairly unfortunate, and something we should aim to improve later.

Done, and filed https://github.com/cockroachdb/cockroach/issues/74799.

> I wonder if we'd want to add a latency assertion for these operations, to make sure we actually fail fast. These can be pretty annoying to test reliably, but it's an important property of the breaker that requests don't appear to hang.

There is a 45s bound on these requests right now, so if they hang the test will fail. If they return in under 45s by chance, through some other mechanism, it's implausible that they would do so with an error from the breaker. But I understand where you're coming from. Don't want to get into that yak shave now though, and adding a (say) five second timeout doesn't really add much in my opinion. A tracing-based verification could make more sense, to check that requests that arrive when the breaker is already tripped catch a log event that we know precedes any place where the request could have blocked. I'll do that in a future PR if you don't mind, though. Added a TODO.

RFAL

erikgrinaker reviewed (approved):
Sounds good, thanks! LGTM!

tbg commented:
bors r=erikgrinaker
